bpf-gpl/tc.c

```diff
@@ -912,15 +514,25 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct __sk_buff *skb,
 						bpf_ntohl(state->ct_result.nat_ip));
 			}
 
-			if (!icmp_skb_get_hdr(skb, &icmp)) {
-				CALI_DEBUG("Ooops, we already passed one such a check!!!\n");
+			/* Related ICMP traffic must be an error response so it should include inner IP
+			 * and 8 bytes as payload. */
+			if (skb_refresh_validate_ptrs(ctx, ICMP_SIZE + sizeof(struct iphdr) + 8)) {
+				CALI_DEBUG("Failed to revalidate packet size\n");
 				goto deny;
 			}
 
-			l3_csum_off += sizeof(*ip_header) + sizeof(*icmp);
-			ip_header = (struct iphdr *)(icmp + 1); /* skip to inner ip */
+			/* Skip past the ICMP header and check the inner IP header. */
+			l3_csum_off += sizeof(*ctx->ip_header) + sizeof(struct icmphdr);
+			ctx->ip_header = (struct iphdr *)(ctx->icmp_header + 1); /* skip to inner ip */
```

Comment:
Does the context definitely not need the outer IP header any more?

Reply:
I believe that this is 1:1 change of what was there originally. From this point on, NAT (iirc) is done on the inner IP header. It might be that with a bigger context and no need to store it on stack, we could have both and make some code simpler. Probably out of the scope of this refactoring.

Reply:
Yes @tomastigera is right, it's a port of the old code; I'll add a warning comment for now; I'm not sure how to rework it to avoid the problem.

Reply:
Thanks, fair enough.  It sounds fine to leave this until there's some actual need for the outer header.